File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
using namespace std;

int n, m;
int dist[100][100];
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

bool isOut(int x, int y) {
    return (x < 0 || x >= n || y < 0 || y >= m);
}

void bfs(vector<vector<int>> maps) {
    queue<pair<int, int>> q;
    
    q.push({0, 0});
    dist[0][0] = 1;
    
    while(!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();
        
        for(int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(isOut(nx, ny) || dist[nx][ny] != -1 || maps[nx][ny] == 0) {
                continue;
            }
            
            q.push({nx, ny});
            dist[nx][ny] = dist[x][y] + 1;
        }
    }
}

int solution(vector<vector<int>> maps)
{
    n = maps.size();
    m = maps[0].size();
    
    memset(dist, -1, sizeof(dist));
    
    bfs(maps);
    
    return dist[n-1][m-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int n, m;
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, -1, 1};
    
    private boolean isOut(int x, int y) {
        return (x < 0 || x >= n || y < 0 || y >= m);
    }
    
    private int bfs(int[][] maps) {
        Queue<int[]> q = new LinkedList<>();
        
        q.offer(new int[]{0, 0});
        
        while(!q.isEmpty()) {
            int[] p = q.poll();
            int x = p[0];
            int y = p[1];
            
            if(x == n - 1 && y == m - 1) {
                return maps[x][y];
            }
            
            for(int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if(isOut(nx, ny) || maps[nx][ny] != 1) {
                    continue;
                }
                
                q.offer(new int[]{nx, ny});
                maps[nx][ny] = maps[x][y] + 1;
            }
        }
        
        return -1;
    }
    
    
    public int solution(int[][] maps) {
        n = maps.length;
        m = maps[0].length;  
        
        return bfs(maps);
    }
}